### ISSUE TITLE
P0 LEAK ROOT CAUSE: Dockerfile val=head-c-100000(train) → 100% byte overlap

### ISSUE THREAD

**Issue by gHashTag** [gHashTag/trios-trainer-igla#60]
## P0 — Train/val 100 KB-prefix overlap (THE leak root cause)

**Discovered:** 2026-04-30 18:35 UTC after the `--ctx` fix shipped in
[gHashTag/trios-railway@69c3467](https://github.com/gHashTag/trios-railway/commit/69c3467f).

### Smoking gun

[`Dockerfile:35-36`](https://github.com/gHashTag/trios-trainer-igla/blob/main/Dockerfile#L35-L36):

```dockerfile
RUN mkdir -p /work/data && \
    curl -sL https://raw.githubusercontent.com/karpathy/char-rnn/master/data/tinyshakespeare/input.txt > /work/data/tiny_shakespeare.txt && \
    head -c 100000 /work/data/tiny_shakespeare.txt > /work/data/tiny_shakespeare_val.txt
```

`tiny_shakespeare_val.txt` is **literally the first 100,000 bytes of
`tiny_shakespeare.txt`**. Train reads the whole file (~1.1 MB). Val
reads bytes 0..100000 of the same file. Train fully contains val.

### Evidence chain

1. Six FIX-VERIFY runs landed at 18:02–18:03 UTC with the `--ctx` fix
   in place (so the trainer actually reached step=4000) and produced
   `final_bpb` in `[0.0006, 0.0015]` — physically impossible for honest
   character-level LM on Shakespeare (irreducible structure ≥ 1 b/B).

| seed | account | dur_s | final_bpb |
|-----:|---------|------:|----------:|
| 42   | acc0 | 171 | 0.0009 |
| 43   | acc1 |  86 | 0.0007 |
| 44   | acc2 | 160 | 0.0010 |
| 1597 | acc3 | 103 | 0.0010 |
| 2584 | acc4 | 103 | 0.0006 |
| 4181 | acc5 | 104 | 0.0015 |

2. The 210 historical `BLITZ-T10H` rows with BPB ∈ [0.0002, 0.0147]
   used the same Docker image, same head-c-100000 split, same overlap.

3. The 6 `gate2_eligible` rows (BPB 1.75–1.82) were **early-stopped at
   step=1000** before the model fully memorised the 100 KB val
   prefix. They're not honest Gate-2 passes — they're snapshots taken
   mid-overfit.

### Hypothesis on why BPB ≠ 0.0 exactly

The model trains on the full 1.1 MB corpus, including bytes 0..100000.
By step 4000 with hidden=384..2048 it has fit the 100 KB prefix to
near-zero error. The remaining 0.0006–0.0015 BPB comes from RNG-driven
dropout / Adam noise at eval time, not honest predictive uncertainty.

### Fix (will be a PR)

Replace the val construction with **last** 100 KB (still wrong — adjacent
bytes leak), **OR** use a deterministic disjoint split with a held-out
seed. Best fix:

```dockerfile
RUN mkdir -p /work/data && \
    curl -sL https://raw.githubusercontent.com/karpathy/char-rnn/master/data/tinyshakespeare/input.txt > /tmp/full.txt && \
    # Train: bytes [0, len-100000); val: bytes [len-100000, len)
    # Strict byte-disjoint split, val seed = 0xDEADBEEF (unused so far)
    SIZE=$(stat -c%s /tmp/full.txt) && \
    head -c $((SIZE - 100000)) /tmp/full.txt > /work/data/tiny_shakespeare.txt && \
    tail -c 100000 /tmp/full.txt > /work/data/tiny_shakespeare_val.txt && \
    rm /tmp/full.txt
```

Plus a **runtime assertion** in `train_loop::load_data` that the val
data is byte-disjoint from train (compare first/last 1 KB hashes).

### Acceptance

- [ ] Dockerfile uses byte-disjoint split (no shared bytes)
- [ ] `train_loop.rs` panics on overlap-detected at load
- [ ] Smoke run on the new image produces `final_bpb >= 1.0` for any
      reasonable seed × hidden combination
- [ ] All 216 leak-tainted rows (210 BLITZ + 6 FIX-VERIFY) are marked
      `last_error='SCARABAEUS-LEAK-CONFIRMED-V2: dockerfile-100k-prefix-overlap'` in NEON
- [ ] CHECKPOINT_POSTMORTEM.md and LEAK_INVESTIGATION.md updated with
      this confirmed root cause

### Related

- [gHashTag/trios-railway@69c3467](https://github.com/gHashTag/trios-railway/commit/69c3467f) — ctx fix that uncovered this
- [gHashTag/trios-railway#101](https://github.com/gHashTag/trios-railway/issues/101) Scarabaeus Engine umbrella
- [openai/parameter-golf#1998](https://github.com/openai/parameter-golf/pull/1998) honest infrastructure submission

phi² + phi⁻² = 3 · R5-honest.
